Add a new configuration option 'config_files'

hetmankp opened the pull request:
This new configuration option allows us to specify one or more configuration files to be used with mypy. Because of how mypy works, only the first found file will be passed to mypy.

Richardk2n committed:
Shifted defaults

Live mode seems to not work properly (for me), so the if/else construct ist changed in a way to use the saved mode whenever possible. Forthermore adapted the regex so it works on my system (windows).

Richardk2n committed:
Merge remote-tracking branch 'upstream/master'

Richardk2n committed:
Find config file and better live mode

Richardk2n committed:
Changed setup and versioning

Richardk2n committed:
Update LICENSE

Richardk2n committed:
Net necessary anymore

Richardk2n committed:
Create python-test.yml

Richardk2n committed:
Update requirements.txt

Richardk2n committed:
Merge branch 'master' of https://github.com/Richardk2n/pyls-mypy

Richardk2n committed:
spdyder compatibility

Richardk2n committed:
formatting experiment

Richardk2n committed:
unnecessary

Richardk2n committed:
Update requirements.txt

Richardk2n committed:
Create python-package.yml

Richardk2n committed:
Delete python-test.yml

Richardk2n committed:
Create python-publish.yml

Richardk2n committed:
update

Richardk2n committed:
Merge branch 'master' of https://github.com/Richardk2n/pyls-mypy

Richardk2n committed:
typo

Richardk2n committed:
typo

Richardk2n committed:
Failed to update variable

Richardk2n committed:
Typing

Richardk2n committed:
formatting

Richardk2n committed:
new version requirement

Richardk2n committed:
Update python-package.yml

Richardk2n committed:
typo

Richardk2n committed:
version bump

Richardk2n committed:
update name. Finished for now

Richardk2n committed:
update pyls

Richardk2n committed:
update requirements

Richardk2n committed:
Update python-package.yml

Richardk2n committed:
Update _version.py

Richardk2n committed:
make configs work on linux

Hi-Angel commented on pyls_mypy/plugin.py:
```diff
@@ -41,17 +44,45 @@ def parse_line(line, document=None):
+            '--follow-imports', 'silent']
+
+    config_files = config.plugin_settings('pyls_mypy').get('config_files', [])
+    config_files = (os.path.expandvars(os.path.expanduser(p.strip()))
```
`p.strip()` call looks unnecessary to me. It would be odd if a user somehow passed trailing whitespace as part of filename configuration. Really: in all my years of programming I had never written a string *(filename or not)* with trailing whitespace inside it. Perhaps, only if you're writing a multiline string, there's a possibility.

And besides, a filename ending with a whitespace is valid too *(even though an inconvenient one)*, so… In this case I'd say, leave it to the user to write configuration filename correctly.

hetmankp replied:
That's a fair point, not sure why I did this, probably too much time in GNU Make.

Hi-Angel commented on pyls_mypy/plugin.py:
```diff
@@ -41,17 +44,45 @@ def parse_line(line, document=None):
+    config_files = config.plugin_settings('pyls_mypy').get('config_files', [])
+    config_files = (os.path.expandvars(os.path.expanduser(p.strip()))
+                    for p in config_files)
+    config_files = [p for p in config_files if p]
```
There's no need to explicitly filter out empty configuration filenames because the `os.path.exists()` calls further below will do that anyway.

hetmankp replied:
I was probably thinking this was more efficient than system calls but it's such a corner case anyway it's probably redundant as you say.

Hi-Angel commented on pyls_mypy/plugin.py:
```diff
@@ -41,17 +44,45 @@ def parse_line(line, document=None):
+        # files in relative paths that are not separated by absolute path
+        # configuration files, will all have an equal chance of being found
+        # in each directory before the search moves up the project tree.
+        config_file = None
```
Whenever you have an identifier bigger than a `i`, `j`, `k`, etc, it's not a good idea to have two identifiers differ in just one character, such as `bars` and `bar`. It may confuse readers. `config_file` may be named, for example, `found_config_file`.

Hi-Angel replied:
> `config_file` may be named, for example, `found_config_file`.

Or, `maybe_config_file` probably would be better, since we got an ADT here *(algebraic data type)*.

hetmankp replied:
That's a fair point. I think `found_config_file` would make more sense. Using conventions from the ML family of languages in Python would probably be confusing for most idiomatic Python programmers. I mean, "Maybe" isn't even the name universally used among languages that include an ADT that represents "some value or no value".

Hi-Angel commented on pyls_mypy/plugin.py:
```diff
@@ -41,17 +44,45 @@ def parse_line(line, document=None):
+
+        if config_file is not None:
+            args += ['--config-file', config_file]
+            break
```
I am a bit confused by this algorithm of choosing a file. IIUC, you simply want to find the first configuration file that exists in the system. There's no need to modify `config_files` variable at all. You simply iterate with `for one_file in config_files:`, and then for the `one_file` you do the checks: whether it's relative, and whether it exists.

For bonus points, if you move the whole check to a separate function, you would eliminate the need to ever mutate `config_file` *(not the `config_files` ;))* variable. It will be as simple as: if file exists, return it; if not, return `None`.

hetmankp replied:
No, that would result in a relative pathed config file higher up the project directory tree in taking precedence over one that's lower, if it appears first in the config file list. This is not ideal as you want all the relative config files highest in the directory tree (i.e. closest to the point of invocation) to be considered first before any further down the tree.

This algorithm ensures that each group of relative pathed files appearing before every group of absolute pathed files will take equal precedence as we scan up the project tree.

Hi-Angel replied:
Oh, I see

Hi-Angel commented on pyls_mypy/plugin.py:
```diff
@@ -41,17 +44,45 @@ def parse_line(line, document=None):
+        if relative:
+            config_files = config_files[len(relative):]
+
+            found = config.find_parents(os.getcwd(), relative)
```
I think `found` variable here should be `paths` or similar. It not immediately obvious what this pyls-specific function does, so naming would have helped a reader.

Hi-Angel commented:
While on it, you also got an excess commit `Merge remote-tracking branch 'upstream/master'`.

hetmankp commented:
> While on it, you also got an excess commit `Merge remote-tracking branch 'upstream/master'`.

@Hi-Angel , I needed to merge from the upstream to keep my own repository up to date. I thought normally upstream maintainers squashed the whole pull request when they merged the commit anyway though if it's different on this project I'd be happy to rework the commit history.

Happy to clean things up further if there's any interest in actually merging this feature.

Hi-Angel commented:
> Happy to clean things up further if there's any interest in actually merging this feature.

Sorry, I must mention, I'm in no way affiliated with this project. I was just interested in the feature you added here, and did a little of review while at it. I have a feeling the project is unmaintained :c

> @Hi-Angel , I needed to merge from the upstream to keep my own repository up to date. I thought normally upstream maintainers squashed the whole pull request when they merged the commit anyway though if it's different on this project I'd be happy to rework the commit history.

I don't know about this project specifically, but what usually people do is `rebase`. That is, you do for example `git rebase upstream/master` *(or just press a button `rebase` on a PR/MR)*, and the code will be rebased without any unnecessary commits appearing.

Hi-Angel commented:
> I don't know about this project specifically, but what usually people do is `rebase`

I must probably elaborate on this. A pull/merge request often has more than one commit "real" commits, i.e. not the "merge" ones. In such case a maintainer can't squash the request, because they'd lose the commit history.

But in your case this may be irrelevant, because you have just one "real" commit, so, you might be right here.

Hi-Angel commented:
Sorry btw, I should say I can't close an opened code discussion, because apparently I chosen the wrong option while sending a review. I got used to GItlab UI, which allows to turn any comment to a discussion, and later close it. Lack of such ability in Github is a nuance I didn't know. Oh, well.

Anyway, I just wanted to say that [this discussion](https://github.com/tomv564/pyls-mypy/issues/48#issuecomment-749803734) might be of interest to you.

haplo committed:
Depend on python-lsp-server

Palantir's python-language-server is now unmaintained.
python-lsp-server is a community-maintained fork.

haplo committed:
Undo black formatting from commit 1489e85d4afa6d34a12

haplo committed:
Revert extra requirements from commit 1489e85d4afa6d34a

haplo committed:
Revert indent on README

haplo committed:
Fix indent in README

Richardk2n committed:
Merge pull request #2 from haplo/python-lsp-server-mypy-ls

Depend on python-lsp-server

Richardk2n committed:
Fixing link formatting

Richardk2n committed:
test 3.9 as well

Richardk2n committed:
version bump

Richardk2n committed:
make pipy accept formatting

haplo committed:
Format codebase with black

Add pre-commit and black pre-commit hook.

Check black formatting in PRs.

Richardk2n committed:
Merge pull request #4 from haplo/black

Format codebase with black

Richardk2n committed:
assure the upload will work

asford committed:
Add dmypy support via mypy.api.run_dmypy

https://github.com/palantir/python-language-server/issues/391

Update plugin flow for non-live-mode dmypy invocation via run_dmypy.

Minor fix to update detect `.mypy.ini` as well as `mypy.ini`.

asford committed:
Fixup README typos from review

Richardk2n committed:
Merge pull request #1 from asford/master

Add dmypy support via mypy.api.run_dmypy

asford committed:
Check dmypy status and kill hung daemons before run

The dmypy daemon process can end up in a hung state for many reasons,
and may persist in a hung state across lsp sessions.

`dmypy run` will block indefinitely the daemon process is unresponsive,
blocking the lsp server and preventing diagnostics.

Add `dmypy status` check before `dmypy run` and kill the daemon if
status is non-zero before diagnostic request. The subsequent run will
then bring up a fresh daemon.

Richardk2n committed:
Merge pull request #10 from asford/dmypy_restart

Check dmypy status and kill hung daemons before run

Richardk2n committed:
add more checks

Richardk2n committed:
fix readme

Richardk2n committed:
Merge pull request #5 from Richardk2n/enforece-uploadability

assure the upload will work

Richardk2n committed:
typo

Richardk2n committed:
version bump

Richardk2n committed:
rename

Richardk2n committed:
typo

Richardk2n committed:
Update test_plugin.py

rchl committed:
Fix wrong configuration being used with multiple workspace folders

When a workspace has multiple folders then the approach of using
one global variable to store the config breaks down as each workspace
folders can have separate configuration and the last one that gets
initialized overwrites the previous one.

Store configuration in a map looked up by workspace folder path.

rchl committed:
run black

Richardk2n committed:
Merge pull request #13 from rchl/fix/workspace-configs

Fix wrong configuration being used with multiple workspace folders

Richardk2n committed:
Merge branch 'master' into rename

Richardk2n committed:
closes #6 closes #7

Also changed the line length and made the config discovery less dependet on string manipulation.
Added deprecation warnings for old namespaces

Richardk2n committed:
update link

Richardk2n committed:
closes #8

Richardk2n committed:
Create mypy.ini

Richardk2n committed:
Merge pull request #11 from Richardk2n/rename

Rename to pylsp prefix

Richardk2n committed:
version bump

Richardk2n committed:
closes #15

jspricke committed:
Switch to unittest mock

jpneverwas committed:
Add option to override command line

Closes #17 and maybe #16.

jpneverwas committed:
Add test fixture to mock out diagnostics cache

meshy committed:
Use subprocess to execute mypy / dmypy

Fixes #17

Richardk2n committed:
Merge pull request #19 from jspricke/mock

Switch to unittest mock

meshy committed:
Replace capture_output shortcut with explicit args

The `capture_output` parameter is a shortcut for setting `stderr` and
`stdout` to `PIPE`, but is not available on Python 3.6. To ensure
compatibility, this uses the longer form.

This approach also has the advantage of not capturing `stdout` when it's
not required on the first call to `dmypy`.

Richardk2n committed:
Merge pull request #23 from meshy/use-subprocess-to-run-mypy

Fix virtualenv detection by using subprocess to execute mypy / dmypy

Richardk2n committed:
version bump

Richardk2n committed:
Version bump again

Something in the CI messed up, no realease was generated -> try again

Richardk2n committed:
test 3.10 as well

Richardk2n committed:
avoid 3.10 beeing read as 3.1

Richardk2n committed:
Update setup.cfg

jpneverwas committed:
fixup! Add option to override command line

jpneverwas committed:
fixup! fixup! Add option to override command line

Richardk2n committed:
fix venvs

Richardk2n committed:
closes #18

Richardk2n committed:
version bump

Richardk2n committed:
Temporary fix to make tests run smoothly

As far as I can tell, mypy does not handle unreachable warnings (which are used in some of the tests) correctly in version 0.920, therefore a version before that is enforced for now.

Richardk2n committed:
version bump

omgitsaheadcrab committed:
chore: add missing config setting and allow mypy 0.920

omgitsaheadcrab committed:
chore: remove mypy version limit from setup.cfg

Richardk2n committed:
fix formatting

The project has a line length of 100. The standard in black is shorter -> different formatting. We have pre-commit hooks for that

Richardk2n committed:
Merge pull request #29 from omgitsaheadcrab/unreachable-code-settings

chore: add missing config setting and allow mypy 0.920

Richardk2n committed:
minor readability changes

Richardk2n committed:
Merge branch 'master' into overrides

Richardk2n committed:
black format

Richardk2n committed:
more expressive name and doc

Richardk2n committed:
Merge pull request #24 from jpneverwas/overrides

Add option to override command line

Richardk2n committed:
version bump

Richardk2n committed:
enforce isort

Richardk2n committed:
update links

Richardk2n committed:
toml support

Richardk2n committed:
minor correction

Richardk2n committed:
use toml

Richardk2n committed:
error in unittest

With C: the code detects the parent directory to be cwd and in cwd there is a config file which therefore gets detected resulting in the unittest failing

Richardk2n committed:
version bump

Richardk2n committed:
Closes #36

Richardk2n committed:
Drop 3.6

Richardk2n committed:
Phase out old names

Richardk2n committed:
version bump

Richardk2n committed:
correct flag

Richardk2n committed:
ALso test on windows

Richardk2n committed:
test action

Richardk2n committed:
action test2

Richardk2n committed:
fix action

Richardk2n committed:
apply defaults

Richardk2n committed:
fix tests

Richardk2n committed:
Exclude non compatible test

Richardk2n committed:
Format

rchl committed:
Expose generic mypy error as diagnostic

rchl committed:
create error or warning depending on process return code

rchl committed:
black reformat

rchl committed:
100 column limit

Richardk2n committed:
Detect all errors

Richardk2n committed:
Merge pull request #40 from rchl/fix/generic-error

Expose generic mypy error as a diagnostic

Richardk2n committed:
version bump

hetmankp committed:
Create new config option 'config_names'

This configuration option allows us to specify additional configuration
file names under which the mypy config could be found.
